subsys/fs/zffs/src/zffs_dir.c

```diff
@@ -0,0 +1,699 @@
+	case ZFFS_TYPE_FILE:
+		*disk.type = ZFFS_TYPE_FILE;
+		sys_put_le32(node_data->file.head, disk.file.head);
+		sys_put_le32(node_data->file.tail, disk.file.tail);
+		sys_put_le32(node_data->file.size, disk.file.size);
+		sys_put_le32(node_data->file.next_id, disk.file.next_id);
+		memcpy(disk.file.name, node_data->name,
+		       strlen(node_data->name));
+
+		rc = zffs_slist_prepend(
+		    zffs, pointer, &dir->list, &node, &disk.dir,
+		    offsetof(struct dir_slist_file_node_disk, name) +
+			strlen(node_data->name));
+		break;
+
+	default:
+		return -EIO;
+	}
+
+	return rc;
+}
+
+int zffs_dir_unlink(struct zffs_data *zffs,
+		     struct zffs_area_pointer *pointer, struct zffs_dir *dir,
+		     const struct zffs_node_data *node_data)
+{
+	sys_snode_t *snode;
+	struct zffs_slist_node node;
+	int rc;
+
+	SYS_SLIST_FOR_EACH_NODE(&zffs->opened, snode)
+	{
+		if (((struct zffs_dir *)snode)->id == node_data->id) {
+			return -EBUSY;
+		}
+	}
+
+	node.id = node_data->id;
+
+	rc = zffs_slist_remove(zffs, pointer, &dir->list, &node);
+	if (rc) {
+		return rc;
+	}
+
+	if (dir->node.id == node_data->id) {
+		dir->node = node;
+	}
+
+	return 0;
+}
+
+static int load_dir(struct zffs_data *zffs,
+		    struct zffs_area_pointer *pointer,
+		    struct zffs_dir_data *data, u32_t len)
+{
+	zffs_disk(u32_t, head);
+	int rc = 0;
+
+	if (len <= sizeof(head)) {
+		return -EIO;
+	}
+
+	rc = zffs_area_read(zffs, pointer, &head, sizeof(head));
+	if (rc) {
+		return rc;
+	}
+	data->head = sys_get_le32(head);
+
+	data->name[len - sizeof(head)] = '\0';
+
+	return zffs_area_read(zffs, pointer, data->name, len - sizeof(head));
+}
+
+static int load_file(struct zffs_data *zffs,
+		     struct zffs_area_pointer *pointer,
+		     struct zffs_file_data *data, u32_t len)
+{
+	struct {
+		zffs_disk(u32_t, head);
+		zffs_disk(u32_t, tail);
+		zffs_disk(u32_t, size);
+		zffs_disk(u32_t, next_id);
+	} disk;
+
+	int rc = 0;
+
+	if (len <= sizeof(disk)) {
+		return -EIO;
+	}
+
+	rc = zffs_area_read(zffs, pointer, &disk, sizeof(disk));
+	if (rc) {
+		return rc;
+	}
+
+	data->head = sys_get_le32(disk.head);
+	data->tail = sys_get_le32(disk.tail);
+	data->size = sys_get_le32(disk.size);
+	data->next_id = sys_get_le32(disk.next_id);
+
+	data->name[len - sizeof(disk)] = '\0';
+
+	return zffs_area_read(zffs, pointer, data->name, len - sizeof(disk));
+}
+
+int zffs_dir_load_node(struct zffs_data *zffs,
+			struct zffs_area_pointer *pointer, u32_t object_size,
+			struct zffs_node_data *data)
+{
+	int rc;
+
+	if (data->id == ZFFS_ROOT_ID) {
+		struct dir_root_disk disk;
+
+		if (object_size != sizeof(disk)) {
+			return -EIO;
+		}
+
+		rc = zffs_area_read(zffs, pointer, &disk, sizeof(disk));
+		if (rc) {
+			return rc;
+		}
+
+		if (*disk.object_type != ZFFS_OBJECT_TYPE_ROOT ||
+		    crc16_ccitt(0, (const void *)&disk, sizeof(disk))) {
+			return -EIO;
+		}
+
+		data->type = ZFFS_TYPE_DIR;
+		data->name[0] = '\0';
+		data->dir.head = sys_get_le32(disk.head);
+
+		return 0;
+	}
+	else {
```

Comment:
the else {} here is unnecessary, it's obvious that if the condition above fails it will end up here (since you always return from the if () {} also)

Reply:
yep